src/aleph/http/http2.clj

```diff
@@ -393,18 +699,704 @@
+
+        aleph-error-handler
+        (fn handle-error [ctx ex]
+          (log/error ex
+                     "Exception caught in HTTP/2 stream client handler"
+                     {:stream-id   h2-stream-id
+                      :raw-stream? raw-stream?})
+          (d/error! response-d ex)
+          (d/success! complete true)
+          (s/close! body-stream)
+
+          (handle-exception* ctx ex false h2-stream))]
+
+    (netty/channel-inbound-handler
+
+      :exception-caught
+      ([_ ctx ex]
+       (log/trace ":exception-caught fired")
+       (aleph-error-handler ctx ex))
+
+      :channel-inactive
+      ([_ ctx]
+       (log/trace "http2/client-handler :channel-inactive fired")
+       (d/success! response-d :aleph/closed)
+       (s/close! body-stream)
+       (d/success! complete true)
+       (.fireChannelInactive ctx))
+
+      :channel-read
+      ([_ ctx msg]
+       (log/trace ":channel-read fired")
+
+       (cond
+         (instance? Http2HeadersFrame msg)
+         (let [headers (.headers ^Http2HeadersFrame msg)
+               body (if raw-stream?
+                      body-stream
+                      (bs/to-input-stream body-stream))
+               ring-resp {:status            (->> headers (.status) (.convertToInt netty/char-seq-val-converter)) ; might be an AsciiString
+                          :headers           (headers->map headers)
+                          :aleph/keep-alive? true           ; not applicable to HTTP/2, but here for compatibility
+                          :aleph/complete    complete
+                          :body              body}]
+           (d/success! response-d ring-resp)
+           (when (.isEndStream ^Http2HeadersFrame msg)
+             (d/success! complete false)
+             (s/close! body-stream)))
+
+         (instance? Http2DataFrame msg)
+         (let [content (.content ^Http2DataFrame msg)]
+           (if raw-stream?
+             (netty/put! (.channel ctx) body-stream content)
+             (do
+               (netty/put! (.channel ctx) body-stream (netty/buf->array content))
+               (.release ^ReferenceCounted msg)))
+           (when (.isEndStream ^Http2DataFrame msg)
+             (d/success! complete false)
+             (s/close! body-stream)))
+
+         :else
+         (.fireChannelRead ctx msg)))
+
+      :channel-read-complete
+      ([_ ctx]
+       (log/trace ":channel-read-complete fired")
+       (netty/flush ctx))
+
+      :user-event-triggered
+      ([_ ctx evt]
+       (handle-user-event-triggered
+         ctx evt stream-go-away-handler reset-stream-handler
+         #(client-handle-shutdown-frame evt h2-stream-id body-stream response-d complete))))))
+
+(defn- max-size-handler
+  "Returns a ChannelHandler that buffers all inbound HEADERS and DATA frames until
+   either (1) end-of-stream appears, or (2) the total DATA frame size exceeds the limit.
+
+   This is for supporting the :max-request-body-size option.
```

Comment:
Does it mean there is no [HttpObjectAggregator](https://netty.io/4.1/api/io/netty/handler/codec/http/HttpObjectAggregator.html) for HTTP2?

Reply:
Nope. There aren't even HttpObjects for H2.

Part of what made this take a lot longer than I'd planned is Netty's H2 code, class hierarchy, etc is 99% separate from the H1 way of doing things.

Reply:
Thanks for the confirmation.